File: chapter12/code12_04.cpp
#include <iostream>
#include <vector>
using namespace std;

void Heap(vector<int> &a, int i, int N)
{
    int child1 = i * 2 + 1;
    if (child1 >= N)
        return;

    if (child1 + 1 < N && a[child1 + 1] > a[child1])
        ++child1;

    if (a[child1] <= a[i])
        return;
    swap(a[i], a[child1]);

    Heap(a, child1, N);
}

void HeapSort(vector<int> &a)
{
    int N = (int)a.size();

    for (int i = N / 2 - 1; i >= 0; --i)
    {
        Heap(a, i, N);
    }
    for (int i = N - 1; i > 0; --i)
    {
        swap(a[0], a[i]);
        Heap(a, 0, i);
    }
}

int main()
{
    int N;
    cin >> N;
    vector<int> a(N);
    for (int i = 0; i < N; ++i)
        cin >> a[i];
    HeapSort(a);
    for (int i = 0; i < a.size(); ++i)
    {
        cout << a[i] << ": ";
    }
    cout << endl;
}

File: chapter12/code12_04.py
def heap(L, i, N):
    child_1 = i*2+1
    if child_1 < N:
        if child_1+1 < N and L[child_1+1] > L[child_1]:
            child_1 += 1
        if L[child_1] > L[i]:
            L[i], L[child_1] = L[child_1], L[i]

        heap(L, child_1, N)


def heap_sort(L):
    n = len(L)

    for i in range(n//2-1, -1, -1):
        heap(L, i, n)

    for i in range(n-1, 0, -1):
        L[0], L[i] = L[i], L[0]
        heap(L, 0, i)


n = int(input())

L = list(map(int, input().split()))

heap_sort(L)

print(L)
